website/blog/2024-04-26-deep-learning-part-2/index.md

```diff
@@ -0,0 +1,287 @@
+
+- **Configuring Auto Logging to log parameters, metrics and checkpoints for all epochs**: Deep learning involves experimenting with various model architectures and hyperparameter settings. Auto logging plays a crucial role in systematically recording these experiments, making it easier to compare different runs and determine which configurations yield the best results. Checkpoints are logged at every epoch, enabling detailed evaluations of all intermediate epochs during the initial exploration phase of the project. However, it is generally not advisable to log all epochs during late-stage development to avoid excessive data writes and latency in the final training stages.
+
+![System Metrics Logged](epoch_logging.png)
+
+The auto-logged checkpoint metrics and model artifacts will be viewable in the MLflow UI as the model trains, as shown below:
+
+![Metrics logged with each epoch](checkpoint_metrics.png)
+
+## The Importance of Logging and Early-stopping
+
+The integration of the Pytorch Lightning `Trainer` callback with MLflow is crucial within this training exercise. The integration allows for comprehensive
+tracking and logging of metrics, parameters, and artifacts during model finetuning without having to explicitly call MLflow logging APIs. Additionally,
+the autologging API allows for modifying the default logging behavior, permitting changes to the logging frequency, allowing for logging to occur at each
+epoch, after a specified number of epochs, or at explicitly defined steps.
+
+### Early stopping
+
+Early stopping is a critical regularization technique in neural network training, designed to assist in preventing overfitting through the act of
+halting training when validation performance plateaus. Pytorch Lightning includes APIs that allow for an easy high-level control of training cessation,
+as demonstrated below.
+
+### Configuring Pytorch Trainer Callback with Early stopping
+
+The example below shows the configuration of the `Trainer` object within `Lightning` to leverage early stopping to prevent overfitting. Once configured, the
+training is executed by calling `fit` on the `Trainer` object. By providing the `EarlyStopping` callback, in conjunction with MLflow's autologging, the
+appropriate number of epochs will be used, logged, and tracked without any additional effort.
+
+```python
+from lightning import Trainer
+from lightning.pytorch.callbacks import EarlyStopping
+
+# Run the training loop.
+trainer = Trainer(
+   callbacks=[
+       EarlyStopping(
+           monitor="Val_F1_Score",
+           min_delta=train_config.min_delta,
+           patience=train_config.patience,
+           verbose=True,
+           mode="max",
+       )
+   ],
+   default_root_dir=train_config.model_checkpoint_dir,
+   fast_dev_run=bool(train_config.debug_mode_sample),
+   max_epochs=train_config.max_epochs,
+   max_time=train_config.max_time,
+   precision="32-true"
+)
+
+trainer.fit(model=nlp_model, datamodule=datamodule)
```

Comment:
This `nlp_model` is not defined above and suddenly appears. I think it's crucial part as it contains pretrained RoBERTa model name and LoRA configurations. Can we add that code snippet?